File: MDVision.IEHR/Controls/Billing/FollowUp/Bill_FollowUpARComments.cs
using MDVision.Business.BCommon;
using MDVision.Business.BLL;
using MDVision.Common.Shared;
using MDVision.Common.Utilities;
using MDVision.Model.Billing.FollowUp;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace MDVision.IEHR.Controls.Billing.FollowUp
{
    public class Bill_FollowUpARComments
    {
        private BLLAdminFollowUp BLLBillingFollowUpObj = null;
        public Bill_FollowUpARComments()
        {
            BLLBillingFollowUpObj = new BLLAdminFollowUp();
        }

        #region Singleton
        private static Bill_FollowUpARComments _obj = null;
        public static Bill_FollowUpARComments Instance()
        {
            if (_obj == null)
                _obj = new Bill_FollowUpARComments();
            return _obj;
        }
        

        #region Private Functions
        private string GetFollowUpComments(Int64 FollowUpCommentId,Int64 VisitId)
        {
            try
            {
                List<FollowUpARComments> objList = BLLBillingFollowUpObj.GetFollowUpComments(FollowUpCommentId,VisitId);
                if (objList.Count>=0)
                {
                    var response = new
                    {
                        status = true,
                        FollowUpCommentInfo = objList,
                        FollowUpCommentInfoCount = objList.Count
                    };
                    return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
                }
                else
                {
                    var response = new
                    {
                        status = false,
                        Message = "Unable to Get Comment List"
                    };
                    return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
                    
                }
                
            }
            catch (Exception ex)
            {
                var response = new
                {
                    status = false,
                    Message = MDVCustomException.HumanReadableMessage(ex.Message),
                };
                return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
            }
        }
        private string SaveFollowUpComments(string fieldsJSON)
        {
            try
            {
                System.Web.Script.Serialization.JavaScriptSerializer ser = new System.Web.Script.Serialization.JavaScriptSerializer();
                var SearchedfieldsJSON = ser.Deserialize<dynamic>(fieldsJSON);

                FollowUpARComments model = new FollowUpARComments();
                model.followUpComments = MDVUtility.ToStr(SearchedfieldsJSON["txtFollowUpArComments"]);
                model.VisitId = MDVUtility.ToLong(SearchedfieldsJSON["hfVisitId"]);
                model.IsFromClaim = MDVUtility.ToBool(SearchedfieldsJSON["hfIsFromClaim"]);
                model.IsDeleted = false;
                #region Database Insertion
                
                BLObject<string> obj = BLLBillingFollowUpObj.InsertNewFollowUpComments(model);
                if (obj.Data != null)
                {
                    var response = new
                    {
                        status = true,
                        Message = Common.AppPrivileges.Save_Message,
                    };
                    return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
                }
                else
                {
                    var response = new
                    {
                        status = false,
                        Message = obj.Message
                    };
                    return Newtonsoft.Json.JsonConvert.SerializeObject(response);
                }
                #endregion
            }
            catch (Exception ex)
            {
                var response = new
                {
                    status = false,
                    Message = MDVCustomException.HumanReadableMessage(ex.Message),
                };
                return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
            }

        }
        private string UpdateFollowUpComments(string fieldsJSON)
        {
            try
            {
                System.Web.Script.Serialization.JavaScriptSerializer ser = new System.Web.Script.Serialization.JavaScriptSerializer();
                var SearchedfieldsJSON = ser.Deserialize<dynamic>(fieldsJSON);

                FollowUpARComments model = new FollowUpARComments();
                model.Id = MDVUtility.ToLong(SearchedfieldsJSON["hfFollowUpCommentId"]);
                model.followUpComments = MDVUtility.ToStr(SearchedfieldsJSON["txtFollowUpArComments"]);
                model.VisitId = MDVUtility.ToLong(SearchedfieldsJSON["hfVisitId"]);
                model.IsFromClaim = MDVUtility.ToBool(SearchedfieldsJSON["hfIsFromClaim"]);
                model.IsDeleted = false;
                #region Database Insertion

                BLObject<string> obj = BLLBillingFollowUpObj.UpdateNewFollowUpComments(model);
                if (obj.Data != null)
                {
                    var response = new
                    {
                        status = true,
                        Message = Common.AppPrivileges.Update_Message,
                    };
                    return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
                }
                else
                {
                    var response = new
                    {
                        status = false,
                        Message = obj.Message
                    };
                    return Newtonsoft.Json.JsonConvert.SerializeObject(response);
                }
                #endregion
            }
            catch (Exception ex)
            {
                var response = new
                {
                    status = false,
                    Message = MDVCustomException.HumanReadableMessage(ex.Message),
                };
                return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
            }

        }

        private string DeleteFollowUpComments(Int64 FollowUpCommentId)
        {
            try
            {
                BLObject<string> obj = BLLBillingFollowUpObj.DeleteNewFollowUpComments(FollowUpCommentId);
                if (obj.Data != null)
                {
                    var response = new
                    {
                        status = true,
                        Message = Common.AppPrivileges.Delete_Message,
                    };
                    return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
                }
                else
                {
                    var response = new
                    {
                        status = false,
                        Message = obj.Message
                    };
                    return Newtonsoft.Json.JsonConvert.SerializeObject(response);
                }
                #endregion
            }
            catch (Exception ex)
            {
                var response = new
                {
                    status = false,
                    Message = MDVCustomException.HumanReadableMessage(ex.Message),
                };
                return (Newtonsoft.Json.JsonConvert.SerializeObject(response));
            }

        }
        #endregion


        #region Service Command Handler

        /// <summary>
        /// Commands the handler.
        /// </summary>
        /// <param name="context">The context.</param>
        public void CommandHandler(HttpContext context)
        {
            string cammandAction = context.Request.QueryString["cammandAction"].ToUpper();

            switch (cammandAction)
            {

                case "SAVE_FOLLOWUP_AR_COMMENTS":
                    {
                        string JSONData = context.Request["Comment_Data"];
                        string strJsonData = SaveFollowUpComments(JSONData);
                        context.Response.ContentType = "text/plain";
                        context.Response.Write(strJsonData);
                    }
                    break;
                case "GET_FOLLOWUP_AR_COMMENTS":
                    {
                        Int64 VisitId = MDVUtility.ToLong(context.Request["VisitId"]);
                        Int64 FollowUpCommentId =MDVUtility.ToLong( context.Request["FollowUpCommentId"]);
                        string strJsonData = GetFollowUpComments(FollowUpCommentId, VisitId);
                        context.Response.ContentType = "text/plain";
                        context.Response.Write(strJsonData);
                    }
                    break;
                case "UPDATE_FOLLOWUP_AR_COMMENTS":
                    {
                        string JSONData = context.Request["Comment_Data"];
                        string strJsonData = UpdateFollowUpComments(JSONData);
                        context.Response.ContentType = "text/plain";
                        context.Response.Write(strJsonData);
                    }
                    break;
                case "DELETE_FOLLOWUP_AR_COMMENTS":
                    {
                        Int64 FollowUpCommentId =MDVUtility.ToLong(context.Request["FollowUpCommentId"]);
                        string strJsonData = DeleteFollowUpComments(FollowUpCommentId);
                        context.Response.ContentType = "text/plain";
                        context.Response.Write(strJsonData);
                    }
                    break;

            }
        }
        #endregion

    }
}

File: MDVision.IEHR/Controls/Billing/FollowUp/Bill_FollowUpARComments.js
Bill_FollowUpARComments = {
    params: [],
    Load: function (params) {
        Bill_FollowUpARComments.params = params;
        Bill_FollowUpARComments.ValidateFollowUpARComments();
        if (Bill_FollowUpARComments.params.VisitId) {
            $('#pnlFollowUpARComments #frmFollowUpARComments #hfVisitId').val(Bill_FollowUpARComments.params.VisitId);
        }
        if (Bill_FollowUpARComments.params.IsFromClaim) {
            $('#pnlFollowUpARComments #frmFollowUpARComments #hfIsFromClaim').val(Bill_FollowUpARComments.params.IsFromClaim);
        }
        if (Bill_FollowUpARComments.params.FollowUpCommentId) {
            Bill_FollowUpARComments.SearchFollowUpComment(Bill_FollowUpARComments.params.FollowUpCommentId);
        }
    },
    SearchFollowUpComment:function(FollowUpCommentId){
        Bill_FollowUpARComments.FollowUpCommentsSearch(FollowUpCommentId,0).done(function (response) {
            if (response.status != false) {
                var obj = response.FollowUpCommentInfo;
                $(response.FollowUpCommentInfo).each(function (i, item) {
                    $('#pnlFollowUpARComments #frmFollowUpARComments #hfVisitId').val(item.VisitId);
                    $('#pnlFollowUpARComments #frmFollowUpARComments #hfIsFromClaim').val(item.IsFromClaim);
                    $('#pnlFollowUpARComments #frmFollowUpARComments #hfFollowUpCommentId').val(item.Id);
                    $('#pnlFollowUpARComments #frmFollowUpARComments #txtFollowUpArComments').val(item.followUpComments);
                });
            }
            else {
                utility.DisplayMessages(response.Message, 2);
            }
        });
    },
    DeleteFollowUpComments: function (FollowUpCommentId) {
        utility.myConfirm('Do you want to delete  this record?', function () {
            Bill_FollowUpARComments.FollowUpCommentsDelete(FollowUpCommentId).done(function (response) {
            if (response.status != false) {
                utility.DisplayMessages(response.Message, 3);
                if (Bill_FollowUpARComments.params.ParentCtrl == "EncounterChargeCapture") {
                    EncounterChargeCapture.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                }
                else if (Bill_FollowUpARComments.params.ParentCtrl == "Bill_FollowUpInsuranceAR_Detail") {
                    Bill_FollowUpInsuranceAR_Detail.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                }

            }
            else {
                utility.DisplayMessages(response.Message, 3);
            }
        });
        }, function () { }, ' Confirm Delete');
    },
    ValidateFollowUpARComments: function () {
        $('#pnlFollowUpARComments #frmFollowUpARComments').bootstrapValidator({
            live: 'disabled',
            message: 'This value is not valid',
            feedbackIcons: {
                valid: 'glyphicon glyphicon-ok',
                invalid: 'glyphicon glyphicon-remove',
                validating: 'glyphicon glyphicon-refresh'
            },
            fields: {
                txtFollowUpArComments: {
                    group: '.col-sm-12',
                    validators: {
                        notEmpty: {
                            message: ''
                        },
                    }
                },
            }
        }).on('success.form.bv', function (e) {
            e.preventDefault();
            Bill_FollowUpARComments.Save();
        });
    },

    Save:function()
    {
       
        var self = $('#pnlFollowUpARComments');
        var myJSON = self.getMyJSONByName();
        if (Bill_FollowUpARComments.params.mode == "Add") {
            Bill_FollowUpARComments.SaveFollowUpComments(myJSON).done(function (response) {
                if (response.status != false) {
                    utility.DisplayMessages(response.Message, 1);
                    Bill_FollowUpARComments.UnLoad();
                    if (Bill_FollowUpARComments.params.ParentCtrl == "EncounterChargeCapture") {
                        EncounterChargeCapture.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                    }
                    else if (Bill_FollowUpARComments.params.ParentCtrl == "Bill_FollowUpInsuranceAR_Detail") {
                        Bill_FollowUpInsuranceAR_Detail.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                    }
                }
            });
        }
        else {
            Bill_FollowUpARComments.UpdateFollowUpComments(myJSON).done(function (response) {
                if (response.status != false) {
                    utility.DisplayMessages(response.Message, 1);
                    Bill_FollowUpARComments.UnLoad();
                    if (Bill_FollowUpARComments.params.ParentCtrl == "EncounterChargeCapture") {
                        EncounterChargeCapture.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                    }
                    else if (Bill_FollowUpARComments.params.ParentCtrl == "Bill_FollowUpInsuranceAR_Detail") {
                        Bill_FollowUpInsuranceAR_Detail.FollowUpCommentLoad(Bill_FollowUpARComments.params.VisitId);
                    }
                }
            });
        }

    },
    SaveFollowUpComments: function (myJSONData) {
        
        var data = "Comment_Data=" + myJSONData;
        return MDVisionService.defaultService(data, "BILLING_FOLLOWUP_COMMENTS", "SAVE_FOLLOWUP_AR_COMMENTS");
    },
    UpdateFollowUpComments: function (myJSONData) {

        var data = "Comment_Data=" + myJSONData;
        return MDVisionService.defaultService(data, "BILLING_FOLLOWUP_COMMENTS", "UPDATE_FOLLOWUP_AR_COMMENTS");
    },
    FollowUpCommentsDelete: function (FollowUpCommentId) {

        var data = "FollowUpCommentId=" + FollowUpCommentId;
        return MDVisionService.defaultService(data, "BILLING_FOLLOWUP_COMMENTS", "DELETE_FOLLOWUP_AR_COMMENTS");
    },
    FollowUpCommentsSearch: function (FollowUpCommentId,VisitId) {

        var data = "FollowUpCommentId=" + FollowUpCommentId + "&VisitId=" + VisitId;
        return MDVisionService.defaultService(data, "BILLING_FOLLOWUP_COMMENTS", "GET_FOLLOWUP_AR_COMMENTS");
    },
    
    UnLoad: function () {
        if (Bill_FollowUpARComments.params.ParentCtrl != "") {
            UnloadActionPan(Bill_FollowUpARComments.params.ParentCtrl);
        }
        else {
            UnloadActionPan();
        }
    },
}